src/simplewallet/simplewallet.cpp

```diff
@@ -565,6 +676,469 @@ bool simple_wallet::print_fee_info(const std::vector<std::string> &args/* = std:
   return true;
 }
 
+bool simple_wallet::prepare_multisig(const std::vector<std::string> &args)
+{
+  if (m_wallet->multisig())
+  {
+    fail_msg_writer() << tr("This wallet is already multisig");
+    return true;
+  }
+  if (m_wallet->watch_only())
+  {
+    fail_msg_writer() << tr("wallet is watch-only and cannot be made multisig");
+    return true;
+  }
+
+  if(m_wallet->get_num_transfer_details())
+  {
+    fail_msg_writer() << tr("This wallet has been used before, please use a new wallet to create a multisig wallet");
+    return true;
+  }
+
+  const auto orig_pwd_container = get_and_verify_password();
+  if(orig_pwd_container == boost::none)
+  {
+    fail_msg_writer() << tr("Your password is incorrect.");
+    return true;
+  }
+
+  std::string multisig_info = m_wallet->get_multisig_info();
+  success_msg_writer() << multisig_info;
+  success_msg_writer() << tr("Send this multisig info to all other participants, then use make_multisig <threshold> <info1> [<info2>...] with others' multisig info");
+  success_msg_writer() << tr("This includes the PRIVATE view key, so needs to be disclosed only to that multisig wallet's participants ");
+  return true;
+}
+
+bool simple_wallet::make_multisig(const std::vector<std::string> &args)
+{
+  if (m_wallet->multisig())
+  {
+    fail_msg_writer() << tr("This wallet is already multisig");
+    return true;
+  }
+  if (m_wallet->watch_only())
+  {
+    fail_msg_writer() << tr("wallet is watch-only and cannot be made multisig");
+    return true;
+  }
+
+  if(m_wallet->get_num_transfer_details())
+  {
+    fail_msg_writer() << tr("This wallet has been used before, please use a new wallet to create a multisig wallet");
```

Comment:
On a second thought, how about allowing the user to create a new separate wallet file only for a specific multisig setup? I mean, the idea is to not force the user to create a new wallet and keep its seed every time he does a multisig transaction. Instead, he could just keep only one seed, and a new multisig-only wallet is generated when `prepare_multisig` is called.

Reply:
I was pondering the same thing on IRC the other day. Why not just roll a new random key when calling `prepare_multisig`. And the purpose of the salt is not clear to me if it's public and constant. Repeating starting key (**b**) would repeat the salted key (**b\***), and if the set-up forces a not-used key for starter (to be fair, the check depends on wallet being synced), then what's the issue with just using the starter key and skip the blinding?

In the MRL paper, the salt is supposed to be like a passphrase chosen by the user, different for each set-up, but that doesn't seem user-friendly to me TBH. Salt should be non-repeating. Maybe using current date/time (say, 1min precision) as salt could work, but user would have to be forced to write it down!

Here's a diagram I drew up while going through the process, this is the current implementation of `prepare_multisig`.
![diag](https://i.imgur.com/6YCpCrM.png)

So, if we roll a new seed, present it to the user, have him confirm that he has written it down before letting him proceed then that ought to be ok, no? If for whatever reason the process is aborted and he wants to continue where he left off, there should be a version of `prepare multisig [old_seed]` to start the process with some previously used seed.

Since we're really making a new wallet, wouldn't it be more natural to do it with `monero-wallet-cli --prepare-multisig [seed]` instead of using an existing one??

Reply:
There's just too many --generate type options, and I think there was another reason I went that way but I forgot it now anyway. If anyone wants to add a --generate option that does the two steps at once, feel free, but this stays as is for now.

Reply:
@JollyMort 
> Why not just roll a new random key when calling `prepare_multisig`.

> So, if we roll a new seed, present it to the user, have him confirm that he has written it down before letting him proceed then that ought to be ok, no? 

My idea is to not generate a new wallet key for each multisig wallet creation, but rather keep using the current normal wallet as-is, and generate a new separate multisig-only wallet _**with the same initial spend secret key (seed)**_. After completing the preparation steps of key exchange with other participants, such a multisig wallet gets its own address and partial secret key(s), which is stored in a separate file (so the original wallet can still be used).

> And the purpose of the salt is not clear to me if it's public and constant.

> In the MRL paper, the salt is supposed to be like a passphrase chosen by the user, different for each set-up, but that doesn't seem user-friendly to me TBH.

I haven't read the MRL paper yet, but AIUI the idea is to allow each participant to hide the connection between the original spend public key `B=b*G` through blinding: `b' = Hs(b || salt); B' = b'*G`. In this sense, salt is just a public constant, just like how the subaddress key is computed using another salt: `D = Hs(a || index || salt)*G + B`.

Reply:
I kinda like that. If it passes the luigi sanity check, I might do that (optional) quick.

Reply:
>After completing the preparation steps of key exchange with other participants, such a multisig wallet gets its own address and partial secret key(s), which is stored in a separate file (so the original wallet can still be used as a normal full wallet).

And you'd have to close the "starter" wallet, and open the newly created one? I like that approach. It's like how `save_watch_only` works. Then we could use any existing wallet as "starter".

>IMHO, though, the benefit of salting (i.e. hide the original wallet address from other participants) doesn't seem too huge to me.

I agree. But. If we go with above approach of using some existing wallet then we could expect users to participate in multiple coalitions with the same "starter" wallet. like:

```
(a, b) -----> (a', b') |----> Coalition 1 (5/5)
                       |----> Coalition 2 (2/3)
                       |----> Coalition 3 (6/7)
```

Now, someone participating in coalition 1&2 could also tell that you are too since your keys would be the same in both. To make things worse, if anyone ever agrees to do the "merging" (exhange secret keys, to cancel multisig and make a normal "master" wallet from it) then he compromises other coalitions with this as well since now some of the secret keys are known by more than 1 person.

Now, how to avoid re-use of (a',b') without too much hassle for the user?

Either we roll a new (a, b), or we make the salt unique for each set-up. It could be current date+hour. Services could make it current timestamp. But the user **must save it**.

Reply:
It seems rather difficult to do a stateful implementation of this; I don't have a good opinion yet.

As far as using a starter wallet, that seems fine to me (with salting still -- I don't see any benefit to using the spend key as is).

Reply:
@JollyMort Yes, the obfuscation from salting becomes moot when two or more individuals form multiple different multisig coalitions. So the benefit of salting seems really questionable, as @luigi1111 also noted.